### ISSUE TITLE
.NET Core support

### ISSUE THREAD

**Issue by matjazmav** [oleg-shilo/cs-script#18]
Will .NET core will be supported?

**Comment by oleg-shilo** [comment]
The short answer is 'Yes it will'.

However, it's not a CS-Script challenge but a  framework one. Right now CS-Script is supported on Linux out of box. This is not because it has some Linux specific functionality but simply because Linux's default .NET implementation MONO has implemented `ICodeCompiler` interface. And as far as I am aware .NET Core is not there yet. 

Saying that CS-Script allows pluggable compilers for the cases when the host system doesn't offer `ICodeCompiler`. Thus CS-Script is using this approach to bring C#6.0 functionality on Windows. Amazingly .NET 4.6, which supports C#6 syntax doesn't implement C# 6 flavor of `ICodeCompiler` so CS-Script is distributed with an alternative compiler (Roslyn) for this. What is even more amazing is that MONO v4.6 in fact does implement C# 6 `ICodeCompiler` out of box. :) Meaning that right now if you are to use cs-script on .NET Core you will need to do an additional configuration step in order to use and additional _codeprovider_ module. Read about it here https://github.com/oleg-shilo/cs-script/wiki/C%236.0-support. 

Anyway I am just waiting for Visual Studio 15 and C#7 to be out because there is a chance that MS team will finally bring the all compilers into the .NET and .NET Core distributions. Then no adjustments will be needed to run CS-Script on .NET Core and no extra user steps will be required.

If it doesn't happen I will need to prepare and start distributing a new version (C#7) of codeprovider.

**Comment by angelog** [comment]
I'm also interested in .Net Core support, as I am currently in the process of porting an application using CS Script. :+1:

**Comment by oleg-shilo** [comment]
I just have finished support for the issue https://github.com/oleg-shilo/cs-script/issues/21

The indirect outcome of the effort is a confirmation that CS-Script works out of box on ASP.NET Core with Roslyn compiler. Though on Windows the back engine is .NET. 

This is rather an encouraging sign that the engine should work on all Core runtimes. Particularly because the indirect signs (like absence of CodeDom interface `ICodeCompiler`) indicate that ASP.NET Core imposes it's own Core interface regardless of the back engine.

I am stressed right now with other CS-Script offsprings so it's unlikely I will be able too look at it before VS15 release. But if you don't want to wait you can give it a try. Just use `CS-Script.lib` NuGet package and Roslyn as a compiling engine. The details are in the issue https://github.com/oleg-shilo/cs-script/issues/21

**Comment by matthiasburger** [comment]
great library for .net framework. i wanted to use it together with my web-application on .net core.
do you have any results so far? else, I'm going to test it for you on .net core this evening and send you some feedback. + if it succeeds I'm going to test it with my web-application, using entity framework and all that stuff.

**Comment by oleg-shilo** [comment]
OK, I have tested the current build against .NET Core (on Mint17). These are the findings:

- The current NuGet package `cs-script.lib` v1.8.1 cannot be integrated into a .Net Core project:
>Package CS-Script.lib 3.18.1 is not compatible with netcoreapp1.1 (.NETCoreApp,Version=v1.1). Package CS-Script.lib 3.18.1 supports: net (.NETFramework,Version=v0.0)

Though this can be fixed as long as CSScriptLibrary.dll is recompiled under .NET Core. In order to accomplish this the `CSScriptLibrary.csproj` will need to be converted into .NET Core `project.json` equivalent. Not sure how complicated this can be but should be possible. May be after the Christmas break :)

- .NET Core is not ready for C# scripting yet. `System.CodeDom.Compiler` not implemented at all. Meaning no native scripting (CodeDom) engine. Mono scripting engine is unlikely due to the Mono team abandoning the effort long time ago. This leaves only Roslyn engine in the picture.

While .NET Core apps can host Roslyn scripting packages after some additional work around (see this [post](https://github.com/dotnet/roslyn/issues/14614) ) it is not fully supported yet. This is what is the state of this according Roslyn team less than 2 months ago from the same post:  
> Note however that .NET Core is not currently supported (see https://github.com/dotnet/roslyn/wiki/Scripting-API-Samples#supported-platforms) hence the experience hasn't been optimized yet. We plan to make it better once we support targeting .NET Core.
____
The bottom line is. CS-Script will need to be adjusted to be ready for the .NET Core. But the actual support can only happen when .NET Core team finalises the effort for Roslyn scripting.

I tagged this issue as an enhancement request. I will also keep it open so anyone who has anything to contribute to the matter can do so.

**Comment by matthiasburger** [comment]
after forking your repository yesterday, I think I got it to run (but was late yesterday) on .net core. I'll do a few more tests after workday and check it in as soon as it works. Hopefully I don't get more errors ;)

**Comment by oleg-shilo** [comment]
I am curious, did you test it against .NET Full or .NET Core?

**Comment by matthiasburger** [comment]
only against .net core. I had to change a lot of things, like you wrote e.g. project.json. I'm not sure whether it'll really work, but I'm on to it.

**Comment by oleg-shilo** [comment]
Great. Keep me posted. 

I also had a quick look and one thing is for sure that all CodeDom.Native, CodeDom.Evaluator and Mono .Evaluator will need to be removed. All reflection based support as well. 
I only tested .NET Core 1.0.1 and it seems to have plenty of API not implemented yet. But one definitely positive outcome is that Roslyn.scripting **can be** hosted successfully. Despite the Roslyn team not claiming that.

I see here an opportunity to make a CS-Script.Core preview. Not a release - neither .NET Core nor Roslyn is ready for that, only a  preview. But..... Christmas is more important. 

Merry Christmas to everyone.

**Comment by oleg-shilo** [comment]
I have managed to have a single complete vertical slice of Roslyn.Evaluator for .NET Core. Had to do plenty of tricks, which may or may not work with the future evolutions of .NET Core.

Started pushing the code to [codebase](https://github.com/oleg-shilo/cs-script/tree/master/Source/CSScript.Core). 

We will see how it goes. If porting the rest of functionality is reasonably smooth then I cannot exclude releasing it as a technology preview effort even before .NET Core stabilizes.

**Comment by danielohirsch** [comment]
Thanks for the work Oleg.  I got that code and tried to use it..  but it doesn't seem to want to automatically reference the assemblies
... Any ideas on how I can make that work?

**Comment by oleg-shilo** [comment]
Grab the latest code. I just did the commit. 

The latest codebase contains the test class that covers currently implemented use cases. It will reference assemblies from the `//css_ref` directives placed in the script code.

Though... do not expect much from this solution. I would rather call it a technology preview effort. 

The .NET Core is still in its infant state when it comes to Reflection. They implemented very little so far. There is no way to discover AppDomain assemblies and auto-reference them. There is no way to resolve name space into assembly name for further and auto-referencing. There is no way to unload already loaded assembly even by the cost of unloading AppDomain. There is no way to emit dynamic delegates. .NET Core just doesn't have these features. May be just yet...

Though while so far there are more gaps than achievements, the most useful/common scripting scenarios are already done. Have a look at `Test.cs` file.

**Comment by oleg-shilo** [comment]
Done: CTP  Release [v3.25.3.0](https://github.com/oleg-shilo/cs-script/releases/tag/v3.25.3.0)

**Comment by graceleecat** [comment]
Hi oleg-shilo，i have read this issue from beginning to the end.
and now it's the end of 2017, does CS-Script have any new change with .NET Core 2.0?  I'm very interested in it :)

**Comment by oleg-shilo** [comment]
Hy there,

I am planing to do another wave update for CTP but to be honest I did not plan to do this until the next year. I am just drowning in the work for CS-Script and WixSharp. Currently the CS-Script for Notepad++ x64 is the highest priority. And WixSharp constant PRs do add extra pressure.

Though I would thing that the CTP should work right away. Do you have some Core 2.0 specific features in mind that you would want to see in CS-Script.Code?

**Comment by graceleecat** [comment]
maybe CodeDom interface ICodeCompiler has been offered in .net core.
NuGet search: System.CodeDom, which is not included in default .net core nuget package.

**Comment by oleg-shilo** [comment]
After packaging recently the latest Rolsyn binaries for CS-Script Notepad++ plugin I noticed a package that indeed looks like CodeDom for Core. I will have a look when I have an opportunity.

Though, I am curious, what would be the advantage to have CodeDom engine integrated for .NET Core.  I see why core application needs to host Roslyn-based scripting interface (CS-Script.Core CTP) but I am not so sure about CodeDom.

**Comment by seertenedos** [comment]
Any chance of posting a dotnet core version of the nugets even if tagged as pre releases? We are in the process of trying to make all our systems compatible with netstandard 2.0 especially all our libs and a number of them rely on the scripting from CS-script

**Comment by oleg-shilo** [comment]
Publishing on NuGet is simple and does not require much time to accomplish. But I am reluctant to do so until I know that what I am publishing works. 

If you can test the (S-Script.Core CTP assembly and can confirm that it works in your environment I will happily publish the  pre-release NuGet package.

**Comment by seertenedos** [comment]
More than happy to test once i know where to find it. It will take a few
days as i am in the middle of converting 20 projects and this is just
affects 3 of them.

**Comment by oleg-shilo** [comment]
The CPT is available since May 2016: https://github.com/oleg-shilo/cs-script/tree/master/Source/.NET%20Core/CSScriptLib

You may also upgrade it and test it for netstandard 2.0. In fact you will be doing me a favor if you do so. 
Once you are happy with the result I will release the package straight away. 

Thank you.

**Comment by seertenedos** [comment]
Oh i thought it was already a nuget just different source like myget or
different name. That will take a bit longer. Will try if i get a chance
after current work.

**Comment by PluginIO** [comment]
I've been following and waiting for this also. 👍  Is this package now available on NuGet?

**Comment by oleg-shilo** [comment]
No this package is only available from GitHub: https://github.com/oleg-shilo/cs-script/tree/master/Source/.NET%20Core/CSScriptLib.

Though the good news is that I finally finished development of the the long overdue Notepad++ CS-Script plugin, which was holding me lately. Thus I will be back on the main CS-Script project in just a few days and... getting .NET Core support out of Beta is the **very next** dev task to do.

**Comment by Yothri** [comment]
As you already posted some releases for .NET Core. I am currently about to implement that. Actually it works, unfortunately Remote Loading is not supported (yet?), but I got another problem. I want to be able to provide Newtonsoft.Json library to the scripts but I cant figure out how to do that. Could you help me?

**Comment by oleg-shilo** [comment]
The news is first: Coding for CSScriptLib for .NET Core 2.0 is done and the release candidate is undergoing testing. Writing the code documentation is also in progress. 

Now about your question.
You can reference the external assembly like this:
```C#
dynamic script = CSScript.Evaluator
                         .ReferenceAssemblyByName("Microsoft.CSharp")
                         .ReferenceAssemblyOf<WebClient>()
                         .ReferenceAssemblyOf<JArray>()
                         .LoadCode(@"using System;
                                     using System.Text;
                                     using System.Net;
                                     using Newtonsoft.Json.Linq;

                                     public class Script
                                     {
                                         public string Run(string url)
                                         {
                                             var client = new WebClient();
                                             client.Headers.Add(""user-agent"", ""anything"");

                                             var releases = client.DownloadString(url);

                                             var report = new StringBuilder();

                                             foreach (dynamic release in JArray.Parse(releases))
                                                 report.AppendLine(release.name.ToString());

                                             return report.ToString();
                                         }
                                        }");

var report = script.Run("https://api.github.com/repos/oleg-shilo/cs-script/releases");
Console.WriteLine(report);
```
In the release I expect to finalize referencing from code so mot likely you will be able to do that:
```C#
dynamic script = CSScript.Evaluator
                         .LoadCode(@"
                            //ss_ref Microsoft.CSharp
                            //css_ref System.Net.WebClient
                            //css_ref Newtonsoft.Json
                            using System;
                            using System.Text;
                            using System.Net;
                            using Newtonsoft.Json.Linq;
```
If CTP does not work for you you can download the hot-RC build from here: https://github.com/oleg-shilo/cs-script/tree/master/Source/.NET%20Core/CSScriptLib/src/CSScriptLib/output

**Comment by Yothri** [comment]
@oleg-shilo That worked flawlessly. I also managed to get it working another way which was more comfortable instead of referencing each explicitly I used Reflection to get all Types loaded in the current executing assembly, then I distinct the types Assembly and referenced them to the Evaluator. That worked pretty good, whereas ReferenceDomainAssemblies is not supported yet :)

Great work, so easy to use actually !

**Comment by oleg-shilo** [comment]
Bingo!!!

Yes this is exactly what the very latest RC does in `ReferenceDomainAssemblies`:
```C#
class RoslynEvaluator
{
    public IEvaluator ReferenceDomainAssemblies(DomainAssemblies assemblies = DomainAssemblies.AllStaticNonGAC)
    {
        foreach (var name in Assembly.GetCallingAssembly().GetReferencedAssemblies())
        {
            var asm = Assembly.Load(name); // the asm is already loaded by the host anyway
            this.ReferenceAssembly(asm);
        }
    }
...
```
Though GetReferencedAssemblies doe not return NuGet packages :(

Too bad .NET Core does not have `AssemblyLoadContext.Default.GetAssemblies()`. Please share your solution if it is better than the one above.

**Comment by Yothri** [comment]
I´ve done it using Reflection:

```
IEvaluator ReferenceDomainAssemblies()
        {
            IEvaluator evaluator = CSScript.Evaluator;
            foreach(var asm in Assembly.GetExecutingAssembly().GetTypes().Select(t => t.Assembly).Distinct())
            {
                evaluator = evaluator.ReferenceAssembly(asm);
            }
            return evaluator;
        }
``` 
That's how it worked for me. I actually didnt use the Distinct() before so that part is pseudocode because I am currently not at my desktop, but it should work.

**Comment by oleg-shilo** [comment]
Hmm, strange. 
First of all I am not sure about your take on getting domain assemblies. Just looking at your solution and I see that it will always return a single assembly:
```C#
var asms = Assembly.GetExecutingAssembly().GetTypes().Select(t => t.Assembly).Distinct().ToArray();
```
You are getting an assembly (executing assembly), getting all the types implemented in this single assembly and after that you are collecting from every type its parent assembly. Which is the very same assembly you have started with.

I tested your code and the test confirmed - it is always a single assembly.

As for "`Newtonsoft.Json` from NuGet" I checked my code  again and indeed it returns Newtonsoft.Json but only if it is already loaded (referenced statically) but not if the host app loads it dynamically (as in my case). Thus it doesn't do the same job as the old `AppDoman.Current.GetAssemblies()`. Ehich is unfortunate...

**Comment by seertenedos** [comment]
Guys has anyone got CS-Script.Core V1.0.0.2 to work in dotnet 4.6.1?  It is netstandard2.0 so it should work but i am getting the below error. We need this as we are trying to slowly port our apps from 4.6.1 to Core2 but are aiming to have them run in both short term plus we have a common lib to do all our scripting stuff used in multi apps.

System.TypeLoadException: Could not load type 'System.Runtime.Loader.AssemblyLoadContext' from assembly 'System.Runtime.Loader, Version=4.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
   at CSScriptLib.RoslynEvaluator.<>c.<GetReferencedAssemblies>b__19_0(PortableExecutableReference r)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at CSScriptLib.RoslynEvaluator.GetReferencedAssemblies()
   at CSScriptLib.RoslynEvaluator.Clone(Boolean copyRefAssemblies)
   at CSScriptLib.CSScript.get_RoslynEvaluator()
   at CSScriptLib.CSScript.get_Evaluator()
   at xxxxx.GetBusinessClassScript[TInstance](String classAsString, Object[] constructorArgs) in xxxxx\OpalScriptManager.cs:line 180
   at xxxxx.GetBusinessFunction[TInstance](String functionScriptBody, String template) in xxxxx\OpalScriptManager.cs:line 327
   at xxxxxTest.Comments_Should_Do_Nothing() in xxxxx\BusinessContentFunctionTest.cs:line 46

tried both 
 TInstance instance = CSScript.Evaluator.ReferenceDomainAssemblies()
                    .LoadCode<TInstance>(classAsString, constructorArgs);
and
CSScript.Evaluator.CreateDelegate<int>( xxx from the given sample

**Comment by seertenedos** [comment]
Also on the loading of Assemblies i tried the line below since it is part of netstandard2.0
System.AppDomain.CurrentDomain.GetAssemblies()

Sadly it errors when adding them.
System.Exception: Current version of Microsoft.CodeAnalysis.Scripting.dll doesn't support referencing assemblies which are not loaded from the file location.
   at CSScriptLib.RoslynEvaluator.ReferenceAssembly(Assembly assembly)

**Comment by seertenedos** [comment]
Further update for loading all assembly's in netstandard2.0 on coreapp2 the following works for me
private IEvaluator ReferenceAppDomainAssemblies()
        {
            IEvaluator evaluator = CSScript.Evaluator;
            foreach (var asm in System.AppDomain.CurrentDomain.GetAssemblies().Distinct())
            {
                if (string.IsNullOrEmpty(asm.Location()))
                {
                    continue;//we need a location
                }
                evaluator = evaluator.ReferenceAssembly(asm);
            }
            return evaluator;
        }

**Comment by seertenedos** [comment]
Issues i mentioned above are resolved in pull request https://github.com/oleg-shilo/cs-script/pull/106

**Comment by Salgat** [comment]
@seertenedos Your suggestion worked great! The only thing I had to change was also add a check for asm.IsDynamic == false (it would throw an exception for me on dynamic assemblies).

**Comment by seertenedos** [comment]
If you turn on preview releases then there is 1.0.3-hotfix release that had
the changes I mentioned plus a few others I found I needed for supporting
netstandard 2 versio on dotnet 4.6.1. be interested to know if that fixes
your issues as well.

**Comment by Methuselah96** [comment]
Is the latest pre-release going to be released as a stable build soon? There are bugs in the latest build that are fixed in the latest pre-release that would be helpful to have in the latest stable build.

**Comment by oleg-shilo** [comment]
Yes it is going to be released withing a few days. To make it in sync with then new VSCode extension release.

Though if I am not mistaken the pre-release changes did not affect .NET Core edition. Can you please name the fixes you are referring to so i can verify if they are already or coming into the next stable release.

**Comment by Methuselah96** [comment]
The latest stable build version `1.0.0.2` was not working with `using` statements in the code being loaded.

**Comment by oleg-shilo** [comment]
I have checked and indeed support for resolving `using` is not in there. Even for the pre-release. This is simply because the `usings` were resolved into GAC assemblies and .NET Core simply does not have GAC.
  
However your post prompted me to look for an alternative approach. Thus I have updated the solution with the resolving `using` against assemblies in the `donet\shared` directory, which is in many ways plays the same role as GAC except it one cannot add custom assemblies ther. Any custom assemblies needs to be added and resolved as NuGet packeges (already supported by cs-script.core). 

This is the solution:
```C#
public static string[] FindGlobalAssembly(string namespaceStr)
{
    // .NET Core does not offer any asm discovery mechanism
    // Thus just check for the candidates in the "global shared" folder

    string shared_dir = Path.GetDirectoryName("".GetType().Assembly.Location);
    return FindLocalAssembly(namespaceStr, shared_dir);
}
```
The stable release is a couple days away.

**Comment by Methuselah96** [comment]
@seertenedos That is indeed to what I was referring.

**Comment by oleg-shilo** [comment]
Done. 
The latest release can be found here: https://github.com/oleg-shilo/cs-script/releases/tag/v3.28.4.0

Or you can download it as NuGet packages.

**Comment by Methuselah96** [comment]
@oleg-shilo I don't see the latest CS-Script.Core (v1.0.1.0) released as a NuGet package.

**Comment by rconde01** [comment]
Can you use cscs with dotnetcore? If so...how?

**Comment by oleg-shilo** [comment]
The latest **CS-Script.Core** class library for hosting the script engine is available at NuGet:
https://www.nuget.org/packages/CS-Script.Core

However the standalone engine executable (cscs.exe) for .NET Core is currently under active development. 

The recent MS announcement of the imminent arrival of _.NET Core 3_ with the support for desktop applications has shifted the intensity of the cscs.exe rework one gear up. The working "almost-aplha" is already in codebase. And the preview release is a matter of a few weeks. 

Though there is a one less exciting news...   Early tests reviled that MS has not ported Roslyn compiler server (`VBCSCompiler.exe`) to .NET Core. Meaning that the initial script compilation/execution all attract a significant startup delay (1-3 seconds). However I expect it to be fixed when .NET Core 3 is released. 

Though only time will tell :)

**Comment by rconde01** [comment]
Thanks for the update...although 2019 doesn't feel too imminent :)

**Comment by oleg-shilo** [comment]
:) I know, 'inevitable' would be a better choice. 
I just couldn't resit. Wishful thinking...

**Comment by agilmore30** [comment]
Hi 

Microsoft announced [.NET Core 3 Preview 1](https://blogs.msdn.microsoft.com/dotnet/2018/12/04/announcing-net-core-3-preview-1-and-open-sourcing-windows-desktop-frameworks/) on December 4. 

You mentioned that there is currently a 1-3 second overhead with the Roslyn compiler. I'm experiencing that, which is a bit of a killer. 

I'm wondering if this time delay issue will go away with .NET Core 3 ?

I'm loading code from a string. I appear to get the delay for every invocation. 

 var code = @"_a class ..._";
 var script =  CSScript.Evaluator.LoadCode< ICScript>(code);

Can I do something different to avoid the delay, e.g. load from a file ?

**Comment by agilmore30** [comment]
Actually my assertion that the 3 second overhead occurs on every invocation was wrong. My  method for timing was invalid. I only get the overhead the first time I call a CSScript.Evaluator.LoadCode method, even across different instances of the Evaluator.

**Comment by oleg-shilo** [comment]
Correct.
The overhead is not even functional from CS-Script point of view. Simply it takes time for CLR for the Roslyn assemblies to load and then to initialize all Roslyn static members. After that it is actually pretty fast.

**Comment by danielohirsch** [comment]
I would orecompile your scripts when the app loads and keep them in a keyed
library to call and execute when you want/need.

**Comment by danielohirsch** [comment]
*precompile

**Comment by agilmore30** [comment]
@oleg-shilo @danielohirsch Thanks for your help guys. Yep, I've found its pretty fast after first execution. Precompiling is a good idea, I'll consider that.
